lib/rules/func-name-matching.js

```diff
@@ -0,0 +1,151 @@
+
+        // module.exports
+        if (pattern.property.type === "Identifier" && pattern.property.name === "exports") {
+            return true;
+        }
+
+        // module["exports"]
+        if (pattern.property.type === "Literal" && pattern.property.value === "exports") {
+            return true;
+        }
+    }
+    return false;
+}
+
+//------------------------------------------------------------------------------
+// Rule Definition
+//------------------------------------------------------------------------------
+
+module.exports = {
+    meta: {
+        docs: {
+            description: "require function names to match the name of the variable or property to which they are assigned",
+            category: "Stylistic Issues",
+            recommended: false
+        },
+
+        schema: [
+            {
+                type: "object",
+                properties: {
+                    includeCommonJSModuleExports: {
+                        type: "boolean"
+                    }
+                },
+                additionalProperties: false
+            }
+        ]
+    },
+
+    create(context) {
+
+        const includeModuleExports = context.options[0] && context.options[0].includeCommonJSModuleExports;
+
+        /**
+         * Reports
+         * @param {ASTNode} node The node to report
+         * @param {string} name The variable or property name
+         * @param {string} funcName The function name
+         * @param {boolean} isProp True if the reported node is a property assignment
+         * @returns {void}
+         */
+        function report(node, name, funcName, isProp) {
+            context.report({
+                node,
+                message: isProp ? "Function name `{{funcName}}` should match property name `{{name}}`"
+                                : "Function name `{{funcName}}` should match variable name `{{name}}`",
+                data: {
+                    name,
+                    funcName
+                }
+            });
+        }
+
+        //--------------------------------------------------------------------------
+        // Public
+        //--------------------------------------------------------------------------
+
+        return {
+
+            VariableDeclarator(node) {
+                if (node.init.type !== "FunctionExpression") {
+                    return;
+                }
+                if (node.init.id && node.id.name !== node.init.id.name) {
+                    report(node, node.id.name, node.init.id.name, false);
+                }
+            },
+
+            AssignmentExpression(node) {
+                if (node.right.type !== "FunctionExpression" || (!includeModuleExports && isModuleExports(node.left))) {
+                    return;
+                }
+                const name = getName(node.left);
+                const isProp = node.left.type === "MemberExpression" ? true : false;
```

Comment:
Maybe we can get simpler by `astUtils`.

```js
                const isProp = node.left.type === "MemberExpression" ? true : false;
                const name = isProp ? astUtils.getStaticPropertyName(node) : node.name;
```

Also, `Literal` might not convert to `Identifier`. I think this rule should ignore such literals.
`obj["foo//bar"] = function bar() {}`